### ISSUE TITLE
Deleting a line with sign breaks the sign's line number

### ISSUE THREAD

**Issue by statiolake** [neovim/neovim#14460]
- `nvim --version`: NVIM v0.5.0-dev+1283-g59eae3b38
<details><summary>Full Neovim version</summary><p>

    NVIM v0.5.0-dev+1283-g59eae3b38
    Build type: RelWithDebInfo
    LuaJIT 2.1.0-beta3
    Compilation: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe /DWIN32 /D_WINDOWS /W3 /MD /Zi /O2 /Ob1 /DNDEBUG /W3 -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DWIN32 -D_WIN32_WINNT=0x0600 -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -ID:/a/neovim/neovim/build/config -ID:/a/neovim/neovim/src -ID:/a/neovim/neovim/nvim-deps/usr/include -ID:/a/neovim/neovim/build/src/nvim/auto -ID:/a/neovim/neovim/build/include
    Compiled by runneradmin@fv-az152-644

    Features: -acl +iconv +tui
    See ":help feature-compile"

        システム vimrc: "$VIM\sysinit.vim"
         省略時の $VIM: "C:/Program Files/nvim/share/nvim"

    Run :checkhealth for more info

</p></details>

- `vim -u DEFAULTS` (version: 8.2 (1-2758)) behaves differently? Yes
<details><summary>Full Vim version</summary><p>

    VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Apr 13 2021 09:28:05)
    MS-Windows 64 ビット コンソール 版
    適用済パッチ: 1-2758
    Compiled by dicen@THUNDERBOLT
    Huge 版 without GUI.  機能の一覧 有効(+)/無効(-)
    +acl                +eval               +multi_byte_ime/dyn -tag_any_white
    +arabic             +ex_extra           +multi_lang         -tcl
    +autocmd            +extra_search       -mzscheme           +termguicolors
    +autochdir          -farsi              -netbeans_intg      +terminal
    +autoservername     +file_in_path       +num64              -termresponse
    -balloon_eval       +find_in_path       +packages           +textobjects
    +balloon_eval_term  +float              +path_extra         +textprop
    -browse             +folding            -perl               -tgetent
    ++builtin_terms     -footer             +persistent_undo    +timers
    +byte_offset        +gettext/dyn        +popupwin           +title
    +channel            -hangul_input       -postscript         -toolbar
    +cindent            +iconv/dyn          +printer            +user_commands
    +clientserver       +insert_expand      +profile            +vartabs
    +clipboard          +ipv6               -python             +vertsplit
    +cmdline_compl      +job                +python3/dyn        +virtualedit
    +cmdline_hist       +jumplist           +quickfix           +visual
    +cmdline_info       +keymap             +reltime            +visualextra
    +comments           +lambda             +rightleft          +viminfo
    +conceal            +langmap            -ruby               +vreplace
    +cryptv             +libcall            +scrollbind         +vtp
    +cscope             +linebreak          +signs              +wildignore
    +cursorbind         +lispindent         +smartindent        +wildmenu
    +cursorshape        +listcmds           +sound              +windows
    -darkmode_w32       +localmap           +spell              +writebackup
    +dialog_con         -lua                +startuptime        -xfontset
    +diff               +menu               +statusline         -xim
    +digraphs           +mksession          -sun_workshop       -xpm_w32
    -dnd                +modify_fname       +syntax             -xterm_save
    -ebcdic             +mouse              +tag_binary         
    +emacs_tags         -mouseshape         -tag_old_static     
          システム vimrc: "$VIM\vimrc"
          ユーザー vimrc: "$HOME\_vimrc"
       第2ユーザー vimrc: "$HOME\vimfiles\vimrc"
       第3ユーザー vimrc: "$VIM\_vimrc"
           ユーザー exrc: "$HOME\_exrc"
        第2ユーザー exrc: "$VIM\_exrc"
      デフォルトファイル: "$VIMRUNTIME\defaults.vim"
    コンパイル: cl -c /W3 /nologo  -I. -Iproto -DHAVE_PATHDEF -DWIN32  -DFEAT_CSCOPE -DFEAT_TERMINAL -DFEAT_SOUND  -DFEAT_JOB_CHANNEL -DFEAT_IPV6       -DWINVER=0x0501 -D_WIN32_WINNT=0x0501 /source-charset:utf-8 /MP -DHAVE_STDINT_H /Ox /GL -DNDEBUG  /Zl /MT /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DDYNAMIC_ICONV -DDYNAMIC_GETTEXT -DFEAT_PYTHON3 -DDYNAMIC_PYTHON3  -DDYNAMIC_PYTHON3_DLL=\"python39.dll\" -DFEAT_HUGE /Fd.\ObjCHKAMD64/ /Zi
    リンク: link  /nologo /opt:ref /LTCG:STATUS oldnames.lib kernel32.lib advapi32.lib shell32.lib gdi32.lib  comdlg32.lib ole32.lib netapi32.lib uuid.lib /machine:AMD64   libcmt.lib  user32.lib      /nodefaultlib:python39.lib    winmm.lib WSock32.lib Ws2_32.lib  /PDB:vim.pdb -debug
</p></details>

- Operating system/version: **Windows** (seems important) 10 19042.928
- Terminal name/version: Reproducible in both Command Prompt and Neovim-qt
- `$TERM`: `vtpcon` (in Command Prompt)

### Steps to reproduce using `nvim -u NORC`

1. Create the following script in the working directory, `repro.vim`.
    ```vim
    set signcolumn=yes
    sign define test texthl=ErrorMsg text=!
    sign place 10 name=test line=5
    ```
1. Prepare some text file `test.txt` which has around 10 lines.
1. Open the Neovim by `nvim -u repro.vim test.txt`.
1. You'll see the sign on the line 5.
    ![image](https://user-images.githubusercontent.com/20490597/116665944-5985c800-a9d5-11eb-8996-7b291feae144.png)
1. Move cursor to line 5 and delete that line.
1. You'll see the sign on the line 5 disappeared. (it's *not* normal)
1. type `:sign place` and you'll see an invalid line number is set to the sign.
    ![image](https://user-images.githubusercontent.com/20490597/116665983-64d8f380-a9d5-11eb-9275-f030eeba83a0.png)
    (Sorry it's Japanese in the screenshot)

I heard (not tested by myself) that this does not reproduce in Linux.

### Actual behaviour

The sign on the removed line disappears and has invalid line number.

### Expected behaviour

The sign should persist in the next line of the deleted line.

This is How Vim and the older Neovim (at least `v0.5.0-dev+1062-gcc1851c9f`) behave.

**Comment by matveyt** [comment]
I assume it is intentional. The relevant commit is f2ed760.

**Comment by statiolake** [comment]
OK, that was surprising difference for me but now I understand Neovim removes signs on a removed line.
Is it intentional too that the removed signs are still in the sign list (`:sign place`) with the negative line number instead of disappearing?

**Comment by matveyt** [comment]
@statiolake  To me it looks like a (harmless?) bug, but this question should probably be addressed to @da-x.

**Comment by statiolake** [comment]
At least vim-signify (https://github.com/mhinz/vim-signify) stops working --- it doesn't expect that the line number can be negative. I can fix the plugin easily by modifying its regex to allow negative values (adding something like `-?`), but whether this fix is appropriate or not depends on the behavior (line number can be negative) is intended or not.

**Comment by mhinz** [comment]
> The sign should persist in the next line of the deleted line.

Why? :) That would mean you would have a sign for an unrelated line.

> At least vim-signify (https://github.com/mhinz/vim-signify) stops working --- it doesn't expect that the line number can be negative.

Hmm, I can't reproduce that with the steps you provided. I played around a bit, but at no time I have a sign with a negative line number.

Also.. with `signcolumn=yes`, like in your repro.vim, the sign shouldn't be removed at all. It should only happen for `auto` or `auto:4` etc.

**Comment by mhinz** [comment]
The fact that you get a line number of `-2147483644` (which is really `INT_MIN + 4`) means that something overflowed. And I suspect it only happens on Windows for some reason.

It would be nice if you could check if it really started happening with https://github.com/neovim/neovim/commit/f2ed7605da45eb79a4f7bb89fb19f680fb5a4927.

**Comment by matveyt** [comment]
@mhinz Confirmed, it's right after f2ed760, Windows build.  `-2147483644` is equal to `0x7fffffff` (`MAXLNUM`) + `5` (sign line number in the example above). The overflow is due to `long` type is 32-bit even on Win64.

**Comment by statiolake** [comment]
> > The sign should persist in the next line of the deleted line.

> Why? :) That would mean you would have a sign for an unrelated line.

Sorry for my poor English... I've never thought the behavior around here was different between Vim and Neovim, so I thought Neovim was wrong just because it didn't behave like Vim. That was the meaning of my "should".

> Hmm, I can't reproduce that with the steps you provided. I played around a bit, but at no time I have a sign with a negative line number.

> The fact that you get a line number of `-2147483644` (which is really `INT_MIN + 4`) means that something overflowed. And I suspect it only happens on Windows for some reason.

I think so too. I heard a Linux user doesn't have this problem and vim-signify is also working as expected.

> Also.. with signcolumn=yes, like in your repro.vim, the sign shouldn't be removed at all. It should only happen for auto or auto:4 etc.

Yes. I was actually very confused yesterday but I couldn't check it with the latest build because yesterday the nightly build once disappeared from the GitHub Release page. Now I downloaded the latest nightly build (v0.5.0-dev+1285) and confirmed Neovim still doesn't remove the sign in this case on Linux. Also confirmed that the latest v0.5.0-dev+1285 had the problem on Windows.

So, In conclusion, in this `signcolumn=yes` case should Neovim behave just like Vim? My current understanding is that the sign removing behavior should  only occur when `signcolumn` is `auto` or `auto:*` or etc., and the behavior described in this issue is completely unintentional.

> It would be nice if you could check if it really started happening with f2ed760.

I agree. I don't know how much I can do with my time and knowledge, but I'll try to bisect the problematic commit.

**Comment by statiolake** [comment]
So the simplest fix is just to preserve the original condition as follows:

```diff
diff --git a/src/nvim/sign.c b/src/nvim/sign.c
index 5c7b497a1..1e12dd60e 100644
--- a/src/nvim/sign.c
+++ b/src/nvim/sign.c
@@ -749,7 +749,7 @@ void sign_mark_adjust(
         }
         xfree(sign);
         continue;
-      } else {
+      } else if (amount != MAXLNUM) {
         new_lnum += amount;
       }
     } else if (sign->se_lnum > line2) {
```

I confirmed this fixes the issue in my environment. Is this enough?
